src/Microsoft.ML.ImageAnalytics/ImagePixelExtractor.cs

```diff
@@ -656,29 +490,222 @@ private VectorType[] ConstructTypes()
+                bool asFloat = ImagePixelExtractingTransformer.Defaults.Convert)
+            {
+                Contracts.CheckNonWhiteSpace(name, nameof(name));
+
+                Name = name;
+                InputColumnName = inputColumnName ?? name;
+                Colors = colors;
+
+                if ((Colors & ColorBits.Alpha) == ColorBits.Alpha) Planes++;
+                if ((Colors & ColorBits.Red) == ColorBits.Red) Planes++;
+                if ((Colors & ColorBits.Green) == ColorBits.Green) Planes++;
+                if ((Colors & ColorBits.Blue) == ColorBits.Blue) Planes++;
+                Contracts.CheckParam(Planes > 0, nameof(colors), "Need to use at least one color plane");
+
+                Interleave = interleave;
+
+                AsFloat = asFloat;
+                if (!AsFloat)
+                {
+                    Offset = ImagePixelExtractingTransformer.Defaults.Offset;
+                    Scale = ImagePixelExtractingTransformer.Defaults.Scale;
+                }
+                else
+                {
+                    Offset = offset;
+                    Scale = scale;
+                }
+                Contracts.CheckParam(FloatUtils.IsFinite(Offset), nameof(offset));
+                Contracts.CheckParam(FloatUtils.IsFiniteNonZero(Scale), nameof(scale));
+            }
+
+            internal ColumnInfo(string name, string inputColumnName, ModelLoadContext ctx)
+            {
+                Contracts.AssertNonEmpty(name);
+                Contracts.AssertNonEmpty(inputColumnName);
+                Contracts.AssertValue(ctx);
+
+                Name = name;
+                InputColumnName = inputColumnName;
+
+                // *** Binary format ***
+                // byte: colors
+                // byte: convert
+                // Float: offset
+                // Float: scale
+                // byte: separateChannels
+                Colors = (ImagePixelExtractingEstimator.ColorBits)ctx.Reader.ReadByte();
+                Contracts.CheckDecode(Colors != 0);
+                Contracts.CheckDecode((Colors & ImagePixelExtractingEstimator.ColorBits.All) == Colors);
+
+                // Count the planes.
+                int planes = (int)Colors;
+                planes = (planes & 0x05) + ((planes >> 1) & 0x05);
+                planes = (planes & 0x03) + ((planes >> 2) & 0x03);
+                Planes = (byte)planes;
+                Contracts.Assert(0 < Planes & Planes <= 4);
+
+                AsFloat = ctx.Reader.ReadBoolByte();
+                Offset = ctx.Reader.ReadFloat();
+                Contracts.CheckDecode(FloatUtils.IsFinite(Offset));
+                Scale = ctx.Reader.ReadFloat();
+                Contracts.CheckDecode(FloatUtils.IsFiniteNonZero(Scale));
+                Contracts.CheckDecode(AsFloat || Offset == 0 && Scale == 1);
+                Interleave = ctx.Reader.ReadBoolByte();
+            }
+
+            public void Save(ModelSaveContext ctx)
```

Comment:
>public [](start = 12, length = 6)

should be internal

Reply:
can't change it atm, as it is public in the interface. Artidoro opened an issue about it. 

---
In reply to: [253243474](https://github.com/dotnet/machinelearning/pull/2372#discussion_r253243474) [](ancestors = 253243474)